server/src/main/java/com/thoughtworks/go/server/persistence/AgentDao.java

```diff
@@ -78,35 +104,271 @@ public Agent agentByUuid(String uuid) {
     public void associateCookie(final AgentIdentifier agentIdentifier, final String cookie) {
         final String uuid = agentIdentifier.getUuid();
         final String key = agentCacheKey(uuid);
-        synchronized (key) {
+
+        List<String> uuids = singletonList(agentIdentifier.getUuid());
+        AgentMutex mutex = agentMutexes.acquire(uuids);
+        synchronized (mutex) {
             transactionTemplate.execute(new TransactionCallbackWithoutResult() {
-                @Override protected void doInTransactionWithoutResult(TransactionStatus status) {
-                    Agent agent = fetchAgentByUuid(uuid);
+                @Override
+                protected void doInTransactionWithoutResult(TransactionStatus status) {
+                    Agent agent = fetchAgentFromDBByUUID(uuid);
                     if (agent == null) {
-                        agent = new Agent(uuid, cookie, agentIdentifier.getHostName(), agentIdentifier.getIpAddress());
+                        agent = new Agent(uuid, agentIdentifier.getHostName(), agentIdentifier.getIpAddress(), cookie);
                     } else {
-                        agent.update(cookie, agentIdentifier.getHostName(), agentIdentifier.getIpAddress());
+                        agent.setFieldValues(cookie, agentIdentifier.getHostName(), agentIdentifier.getIpAddress());
                     }
                     getHibernateTemplate().saveOrUpdate(agent);
+
+                    final Agent updatedAgent = agent;
+                    registerCommitCallback(() -> clearCacheAndNotifyAgentEntityChangeListeners(key, updatedAgent));
+                }
+            });
+            agentMutexes.release(uuids, mutex);
+        }
+    }
+
+    String agentCacheKey(String uuid) {
+        return (AgentDao.class.getName() + "_agent_" + uuid).intern();
+    }
+
+    public List<Agent> getAllAgents() {
+        return ((List<Agent>) transactionTemplate.execute((TransactionCallback) transactionStatus -> {
+            Query query = sessionFactory.getCurrentSession().createQuery("FROM Agent where deleted = false");
+            query.setCacheable(true);
+            try {
+                return query.list();
+            } catch (Exception e) {
+                return emptyList();
```

Comment:
We are silently failing here, should we log or add a server health message?

Reply:
If this request fails for some reason the agents in db will never be loaded right?

Reply:
Incorporate following changes,

`AgentService.getAllAgents()` is called only from one place where the Agent Service is initialized.

So when there is an error while loading all agents from DB, throwing an exception in `AgentDao.getAllAgents()` without logging - as the application initialization code which calls `AgentService.initialize()` logs the exception. Hence to avoid double logging, not logging the error in `AgentDao`.